tools/latency_logging/print_timestamps.py

```diff
@@ -0,0 +1,166 @@
+
+r = Route("9f583b1d93915c31|2022-03-28--17-25-12") 
+lr = LogReader(r.log_paths()[0])
+
+services = ['camerad', 'modeld', 'plannerd', 'controlsd', 'boardd']
+
+msgq_to_service = {}
+msgq_to_service['roadCameraState'] = "camerad"
+msgq_to_service['wideRoadCameraState'] = "camerad"
+msgq_to_service['modelV2'] = "modeld"
+msgq_to_service['lateralPlan'] = "plannerd"
+msgq_to_service['longitudinalPlan'] = "plannerd"
+msgq_to_service['sendcan'] = "controlsd"
+
+service_to_durations = defaultdict(list)
+service_to_durations['camerad'] = ['timestampSof', 'processingTime']
+service_to_durations['modeld'] = ['modelExecutionTime', 'gpuExecutionTime']
+service_to_durations['plannerd'] = ["solverExecutionTime"]
+
+# Build translation dict
+translationdict = {}
+for msg in lr:
+  if msg.which() == "logMessage" and "translation" in msg.logMessage:
+    msg = msg.logMessage.replace("'", '"').replace('"{', "{").replace('}"', "}").replace("\\", '')
+    jmsg = json.loads(msg)['msg']['timestampExtra']
+    logMonoTime = jmsg['info']['from']
+    frame_id = jmsg['info']['to']
+    translationdict[float(logMonoTime)] = int(frame_id)
+
+failed_transl = 0
+def translate(logMono):
+    logMono = float(logMono)
+    if logMono in translationdict:
+        return translationdict[logMono]
+    else:
+        global failed_transl
+        failed_transl+=1
+        return -1
+
+# Build service blocks
+service_blocks = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+skip_frames = []
+durations = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+for msg in lr:
+    if msg.which() == "logMessage" and ("smInfo" in msg.logMessage or "timestampExtra" in msg.logMessage):
+        msg = msg.logMessage.replace("'", '"').replace('"{', "{").replace('}"', "}").replace("\\", '')
+        jmsg = json.loads(msg)['msg']['timestampExtra']
+        #make more robust than strings
+        if "smInfo" in msg:
+            pubTime = jmsg['info']['logMonoTime']
+            rcvTime =jmsg['info']['rcvTime']*1e9
+            msg_name = jmsg['info']['msg_name']
+            service = msgq_to_service[msg_name]
+            smInfo = jmsg['info']['smInfo']
+            frame_id = translate(pubTime) if msg_name == 'sendcan' else smInfo['frameId']
+            if msg_name == 'modelV2':
+                if smInfo['frameIdExtra'] != frame_id:
+                    skip_frames.append(frame_id)
+            for duration in service_to_durations[service]:
+                if duration == 'timestampSof':
+                    service_blocks[int(frame_id)][service]["Start"].append(float(smInfo[duration]))
+                else:
+                    durations[int(frame_id)][service][duration].append(smInfo[duration]*1e3)
+            service_blocks[frame_id][service]["End"].append(float(pubTime))
+            next_service = services[services.index(service)+1]
+            service_blocks[int(frame_id)][next_service]["Start"].append( float(rcvTime))
+        elif "Pipeline end" in msg:
```

Comment:
why is this needed?

Reply:
Because sendcan is sent with usb. this is not logged.